### ISSUE TITLE
return_types() confused after running many tests

### ISSUE THREAD

**Issue by nalimilan** [JuliaLang/julia#21740]
I have a branch of CategoricalArrays [here](https://github.com/JuliaData/CategoricalArrays.jl/tree/nl/recodebug) where a complex call to `Base.return_types` returns `Array{String, 1}` when called immediately after loading the package, but `Union{Array{_,1} where _, NullableArrays.NullableArray{_,1} where _}` when called after running lots of tests. If `Base.return_types` is called immediately, then it succeeds later, even after running the tests.

I couldn't find a simple reproducer for now, but if you checkout that branch you can reproduce it easily:
```julia
using CategoricalArrays
# Uncomment to trigger bug (the tests will fail)
# include("test/recode.jl")
Base.return_types(recode, Tuple{Vector{Int}, Void, Pair{Int, String}, Pair{UnitRange{Int}, String}, Pair{Vector{Int}, String}, Pair{UnitRange{Int}, String}})
```

**Comment by yuyichao** [comment]
The return value of `return_types` can certainly change after you've defined new methods.

**Comment by nalimilan** [comment]
Yes, but the `include` call does not define any new methods. I just tests `recode` for many different cases.

**Comment by vtjnash** [comment]
Was this fixed by either of #21677 or #21933?

**Comment by nalimilan** [comment]
Indeed, tests now pass on master, so that was probably #21677 (since #21933 isn't merged yet).

**Comment by nalimilan** [comment]
It's also fixed in 0.6.0-rc3.0, which is really cool.
